iteration of branching convention (#6177)
onflow/flow-go: DEVELOPMENT.md

# Development

Please carefully read our [Coding Conventions](./CodingConventions.md).

## Branching Convention

**This convention is _binding_ for all PR requests in this repo!**
Deviating from this convention has a significant risk of compromising mainnet. 
If you feel that the branching convention could be improved, please prepare a concrete proposal and add an agenda item for
a [Core Protocol Working Group](https://github.com/onflow/Flow-Working-Groups/tree/main/core_protocol_working_group) meeting.
 
### Overview

> On a conceptual level, we maintain three categories of branches: 

Reviewer: ?
```suggestion
On a conceptual level, we maintain four categories of branches: 
```

> * The **deployment branches** version software was/is/should be running on testnet and mainnet. 

Reviewer: ```suggestion
* **deployment branches** contain software versions that was/is/should be running on testnet and mainnet. 
```

  the minor version increases and we create a new deployment branch from `master`.

#### Naming Convention:
The naming convention is `v0.minor`, with the major version currently pinned to `0`.
The minor version increments at every Height-Coordinated-Upgrade [HCU] or spork (points of breaking downwards compatibility).
An example is the branch `v0.33`. Extensions of the branch names are allowed but optional, for example including key differences
such as cryptographic stack (e.g. `v0.33-relic`), storage implementation (e.g. `v0.33-storehouse`), etc.  Patch versions (e.g. `0.33.0`) are usually only included in tags and builds. 
 

### Feature branches
- During development, all features should live on a feature branch.
- For small features, this will be a simple working branch. These branches have the naming scheme `<contributor>/<issue #>-<short description>`, for example `kan/123-fix-known-issue`
- For larger features, this may be a shared feature branch with other team members. Feature branches have the naming scheme `feature/<name>`.

### The `spork` branch

> The spork branch was specifically introduced for managing spork-level changes. Sporks are very far apart (targeted at 12 months intervals) and for of the time in between sporks, 
> the `spork` branch is not needed. HCU-compatible features are merged to `master` and HCU-incompatible features live in their own feature branches until shortly before the spork.

Reviewer: ```suggestion
The spork branch was specifically introduced for managing spork-level changes. Sporks are very far apart (targeted at 12 months intervals), so the `spork` branch is not needed for the time in between sporks.
HCU-compatible features are merged to `master` and HCU-incompatible features live in their own feature branches until shortly before the spork.
```

In an ideal world, where we finalize the spork date a few months ahead of time and _always_ stick to it, the `spork` branch would not be needed at all. 

> Unfortunately, despite our best efforts it has happened that needed to postpone the spork just a few weeks before it was supposed to happen.

Reviewer: ```suggestion
Despite our best efforts, occasionally circumstances arise that require us to postpone the spork just a few weeks before it was supposed to happen.
```

# Development

Please carefully read our [Coding Conventions](./CodingConventions.md).

## Branching Convention

**This convention is _binding_ for all PR requests in this repo!**

> Deviating from this convention has a significant risk of compromising mainnet. 

Reviewer 2: Is it worth mentioning that if a deviation already occurred, or if a deviation has to be made due to to the current scenario we are in (unknown future scenario), the deviation should be clearly communicated to a specific channel?

Author: thanks. Addressed your suggestion in commit https://github.com/onflow/flow-go/pull/6177/commits/c0d09091b00ed01d17de7a1063714e1cfc011f93. I didn't mention a specific channel, because that might change to quickly. Nevertheless, I think its important to mention the responsibility for clear communication.

If you feel that the branching convention could be improved, please prepare a concrete proposal and add an agenda item for
a [Core Protocol Working Group](https://github.com/onflow/Flow-Working-Groups/tree/main/core_protocol_working_group) meeting.
 
### Overview
On a conceptual level, we maintain three categories of branches: 

> * The **deployment branches** version software is was/is/shoud be running on testnet and mainnet. 

Reviewer 3: ```suggestion
* The **deployment branches** version software was/is/should be running on testnet and mainnet. 
```

* The `spork` branch, will be created in preparation of a new spork from the most recent `master`. Feature branches holding HCU-**in**compatible upgrades will be
  consolidated on this `spork` branch. The spork branch is intended to be very short-lived and only exist for a few weeks right before the testnet spork.
  When the Flow network has committed to the spork on a governance level (at the moment, largely coordinated by the Flow Foundation), the `spork` branch will be
  merged back to `master`.

We are purposefully continuing the usage of the 'spork' terminology here, to describe a "severely breaking change".
Specifically, spork describes a deployment of new node software, which requires the entire network to be stopped and rebooted.
We only carry over a snapshot of the execution state but fully re-initialize the protocol state. Thereby, major upgrades of the protocol,
incl. migrations of the entire execution state to a new formats become reasonably straight forward.
Additional context on HCU vs sporks can be found [here](https://developers.flow.com/networks/node-ops/node-operation/hcu#hcu-versus-spork).   

> <!---
> TEMPORARY to show the image, will be replaced with commented-out version below before merge 
> -->
> 
> <img src='https://github.com/onflow/flow-go/blob/ff8882b2fcef5b7dfa86ac97cde407cc7b9ae3d5/docs/flow-go_branching-convention.png' width='350'>
> 
> <!---
> <img src='https://github.com/onflow/flow-go/blob/master/docs/BranchingConvention.png' width='350'>
> -->

Reviewer 3: ```suggestion

<img src='./docs/flow-go_branching-convention.png' width='350'>

```

I believe both GitHub and most IDEs will interpret relative paths, so we can link to whatever file is at this path in the current branch.

#### Motivations
* Based on the decentralized nature of blockchain networks, continuous deployment is practically intractable (without compromising decentralization).
Therefore, longer wait times for features until their deployment must be accounted for in the development process. This convention has been incrementally
developed and refined based on the multi-year learnings of managing the deployment of the flow network.
* Deployment branches, `master` containing a runnable snapshot of the most recent development state and feature branches are intuitively exactly what you
  would expect without further flow-specific context. We just _extend_ the common convention by a few edge cases accounting for the upgrade-constraints of a blockchain network:

>   * On a specific deployment branch, there can only be non-braking changes. This is to prevent accidents where we roll out new node

Reviewer 3: ```suggestion
  * On a specific deployment branch, there can only be non-breaking changes. This is to prevent accidents where we roll out new node
```

>     software incrementally to one node after another, but then discover later that there is some specific case where the two versions don't work together and mainnet halts. 

Reviewer 4: unnecessary to use mainnet as example here.
```suggestion
    software incrementally to one node after another, but then discover later that there is some specific case where the two versions don't work together and network halts. 
```

  If we collected the HCU-incompatible features in a single 'spork' branch, we would need to merge master into that. That means an engineer would need to know all HCU-incompatible features
  currently in the pipeline to resolve merge conflicts. In our model where teams develop features in parallel, it has proven to be efficient for each team to maintain their own HCU-incompatible feature
  branches and regularly merge the evolving `master` into them. As feature branches stay aligned with `master`, the last remaining step of merging all the feature branches together is usually not a big lift.      
* During the limited lifetime of the `spork` branch, the evolution of `master` should be very small. Generally, the engineering team is more focused on testing, a significant portion of features will have
  completed their development in time for the spork and new features start in their own branches anyway.      

### The [master](https://github.com/onflow/flow-go/tree/master) branch

The `master` branch is intended to only contain features for the _immediately_ upcoming release.
This is under the assumption that we already committed to the type of upgrade (HCU vs Spork).

> **Generally only HCU-compatible changes are allowed to be mered to `master`**. It is the responsibilities of the developers to provide conclusive evidence why

Reviewer 2: ```suggestion
**Generally only HCU-compatible changes are allowed to be merged to `master`**. It is the responsibilities of the developers to provide conclusive evidence why
```

their change can be deployed via HCU. For fully downwards-compatible changes, this explanation can be omitted. Otherwise, please provide a brief summary in your PR
targeting master as of why the code changes are HCU-compatible (for example "only affects transaction execution and verification"). More details are provided in the 
[breaking change classification](#breaking-change-classifications) section below.

### Deployment branches

- For every HCU and spork, a new deployment branch will be created from master. This branch will be tagged and used to update testnet and then mainnet.
- **Only non-breaking changes** can be committed to a deployment branch, such that nodes running the older and newer version on this branch can be mixed.
  In other words, a single deployment branch spans _all_ patches that are protocol-compatible without an HCU. 

> - Each time there is a breacking-change in the protocol (see [breaking change classification](#breaking-change-classifications) below),

Reviewer 3: ```suggestion
- Each time there is a breaking-change in the protocol (see [breaking change classification](#breaking-change-classifications) below),
```
